ccm_web/server/src/api/api.service.ts

```diff
@@ -37,66 +41,92 @@ export class APIService {
     const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
-    try {
-      const endpoint = `courses/${courseId}`
-      const method = 'PUT'
-      const requestBody = { course: { name: newName, course_code: newName } }
-      logger.debug(
-        `Sending request to Canvas - Endpoint: ${endpoint}; Method: ${method}; Body: ${JSON.stringify(requestBody)}`
-      )
-      const response = await requestor.requestUrl<CanvasCourse>(endpoint, method, requestBody)
-      logger.debug(`Received response with status code ${response.statusCode}`)
-      const course = response.body
-      return { id: course.id, name: course.name }
-    } catch (error) {
-      const errResponse = handleAPIError(error, newName)
-      return { statusCode: errResponse.canvasStatusCode, errors: [errResponse] }
-    }
+    const courseHandler = new CourseApiHandler(requestor, courseId)
+    return await courseHandler.putCourse({ name: newName, course_code: newName })
   }
 
-  async createSections (user: User, course: number, sections: string[]): Promise<CanvasCourseSection[] | APIErrorData> {
+  async createSections (user: User, courseId: number, sections: string[]): Promise<CanvasCourseSection[] | APIErrorData> {
     const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
-    const createSectionsApiHandler = new CreateSectionApiHandler(requestor, sections, course)
-    return await createSectionsApiHandler.createSections()
+    const courseHandler = new CourseApiHandler(requestor, courseId)
+    return await courseHandler.createSections(sections)
   }
 
   async enrollSectionUsers (user: User, sectionId: number, sectionUsers: SectionUserDto[]): Promise<CanvasEnrollment[] | APIErrorData> {
     const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
-    const enrollmentHandler = new EnrollSectionUsersApiHandler(requestor, sectionUsers, sectionId)
-    return await enrollmentHandler.enrollUsers()
+    const sectionHandler = new SectionApiHandler(requestor, sectionId)
+    return await sectionHandler.enrollUsers(sectionUsers)
+  }
+
+  async mergeSections (user: User, targetCourseId: number, sectionIds: number[]): Promise<CanvasCourseSectionBase[] | APIErrorData> {
+    const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
+    const courseHandler = new CourseApiHandler(requestor, targetCourseId)
+    return await courseHandler.mergeSections(sectionIds)
+  }
+
+  async unmergeSections (user: User, sectionIds: number[]): Promise<CanvasCourseSectionBase[] | APIErrorData> {
+    const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
+    const NS_PER_SEC = BigInt(1e9)
+    const start = process.hrtime.bigint()
+    const apiPromises = sectionIds.map(async (si) => {
+      const sectionHandler = new SectionApiHandler(requestor, si)
```

Comment:
What is the thinking behind putting the merge/unmerge sections in separate handlers course/section. But in the course handler you are calling the section handler for unmerging it. I felt for the API with course_id dependency seems to be in course handler and things with sections dependency seems to be section. 

I think for future API call, If you document the intent that it would be easier for follow along the similar pattern

Reply:
This is a good question. Let me try to explain why I did it this way, and then maybe you can give some advice on how to document it for future use.

I put `mergeSection` and `unmergeSection` in the `SectionHandler` because the routes in Canvas are section-oriented, meaning they use `sections/x` as a base, like the other section calls (`sections/${this.sectionId}/crosslist/${targetCourseId}` and `sections/${this.sectionId}/crosslist`).

In the code you commented on, I chose to only use `SectionApiHandler`s for `unmergeSections` because there isn't a need to identify a course in that operations (see above route). For `mergeSections`, I could have done something similar and not used `CourseApiHandler` at all, but I thought it might be convenient to have a method on that class that would merge multiple sections to one course. I used `SectionApiHandler` in that method since the operation for merging was already defined.

Hopefully that helps explain my logic? If you'd prefer, I can move the `CourseApiHandler.mergeSections` code to `APIService.mergeSections`, I'm not super attached to one implementation over another.

Reply:
> I can move the CourseApiHandler.mergeSections code to APIService.mergeSections

Don't need to move anything. Document it saying that based on the canvas API routing respective API call are placed in respective handler, that's should be good enough for understanding.

Reply:
Cool! There is more we could do with the handler refactoring. We could probably add an intermediary handler class for accounts (`AccountApiHandler`) and a non-admin handler (instructor/student), but I didn't see a huge benefit to those (at least not yet). We could also add abstract base classes, do other OOPsy stuff, but we can explore all this later if we wanted to build a more robust API. Also, GraphQL could change this significantly, whenever that rolls around. I will open a separate issue to capture the ideas here, but it's definitely Icebox material.